### ISSUE TITLE
feature req- ability to set max concurrent queries against a single catalog

### ISSUE THREAD

**Issue by tooptoop4** [prestosql/presto#3103]
presto can connect to many different catalogs ie sqlserver, hive, mysql.etc. While hive might be fine with 100s of concurrent connections to it, some small sqlserver might not. If there are 80 concurrent queries (I know each query only makes one connection to the legacy db but there could be many different queries originating from presto running at same time) against a sqlserver it could overload the 'legacy source system'. Problem with resource groups is they are too coarse (ie they just limit presto wide, not distinguishing between number of queries on hive vs sqlserver). This is about either having new config to limit parallel queries to a certain catalog or adding another dimension of catalog to resource groups.

from slack:

t  7 days ago
when using presto against jdbc sources like sql server,mysql is there some setting to control max parallel connections to those jdbc sources? to avoid overwhelming them

tomn  7 days ago
t, from what I understand, those are serial connectors so they only open 1 connection. There are a limited number of parallel connectors. (hive is the major one)

t  7 days ago
I'm asking about presto server wide, so if there were 100 users all doing presto-mysql queries at once

tomn  7 days ago
Ah, that's a good question. I could see a very busy presto cluster overwhelming a rdbms database. I'll let others answer but as presto is used more and more for federation, this would be a good feature to queue queries that come up if max connections are hit on a certain connector.
:+1:
1

david  7 days ago
You can use resource groups to control the number of concurrent Presto queries. There is no way to control the limit for specific connectors, however.

**Comment by tooptoop4** [comment]
any idea on this? @takezoe

**Comment by takezoe** [comment]
Yeah. Actually, we faced the same issue which hits the max connections in PostgreSQL side. So, I requested to add retry fashion to connecting JDBC data sources in https://github.com/prestosql/presto/pull/3125. Setting max connections lower in PostgreSQL side and leveraging retrying on Presto side could protect PostgreSQL and make Presto queries more robust.

However, this issue can happen on any other data sources, so it would be nice if Presto has a general mechanism to throttle the number of concurrent queries on each catalog. A challenge is that since each presto-worker can connect to data sources, cluster-level limitation would be necessary, not instance-level.

**Comment by tooptoop4** [comment]
BigQuery connector would also benefit from this feature as it has max 100 concurrent query limit (https://cloud.google.com/bigquery/quotas)
